src/theory/arith/nl/nl_model.cpp

```diff
@@ -271,18 +269,25 @@ bool NlModel::checkModel(const std::vector<Node>& assertions,
 
 bool NlModel::addSubstitution(TNode v, TNode s)
 {
-  // should not substitute the same variable twice
   Trace("nl-ext-model") << "* check model substitution : " << v << " -> " << s
                         << std::endl;
+  Assert(getSubstitutedForm(s) == s)
+      << "Added a substitution whose range is not in substituted form " << s;
+  // should not substitute the same variable twice
   // should not set exact bound more than once
   if (d_substitutions.contains(v))
   {
     Node cur = d_substitutions.getSubs(v);
     if (cur != s)
     {
-      Trace("nl-ext-model") << "...ERROR: already has value: " << cur << std::endl;
-      // this should never happen since substitutions should be applied eagerly
-      Assert(false);
+      Trace("nl-ext-model")
+          << "...warning: already has value: " << cur << std::endl;
+      // We set two different substitutions for a variable v. If both are
+      // constant, then we throw an error. Otherwise, we ignore the newer
+      // substitution and return false here.
+      Assert(!cur.isConst() || !s.isConst())
+          << "Conflicting exact bounds given for a variable (" << cur << " and "
+          << s << ") for " << v;
```

Comment:
Does this mean the newer substitutions is never used anywhere? If so, I'd argue we should try to use the simpler substitution instead of the first one (e.g. prefer a constant over a RANs or alike).

Reply:
Ah I see. Okay then.